Main.hs

```diff
@@ -4,26 +4,85 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Number   Double
+  | Sqrt     Expr
+  | Add      Expr Expr
+  | Subtract Expr Expr
+  | Multiply Expr Expr
+  | Divide   Expr Expr
+  | Power    Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Number x)     = show x
+  show (Sqrt x)       = "sqrt(" ++ show x ++ ")"
+  show (Add x y)      = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Subtract x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Multiply x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Divide x y)   = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Power x y)    = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (Number x)       == (Number y)       = x == y
+  (Sqrt x)         == (Sqrt y)         = x == y
+  (Add x1 y1)      == (Add x2 y2)      = x1 == x2 && y1 == y2
+  (Subtract x1 y1) == (Subtract x2 y2) = x1 == x2 && y1 == y2
+  (Multiply x1 y1) == (Multiply x2 y2) = x1 == x2 && y1 == y2
+  (Divide x1 y1)   == (Divide x2 y2)   = x1 == x2 && y1 == y2
+  (Power x1 y1)    == (Power x2 y2)    = x1 == x2 && y1 == y2
+  _ == _ = False
 
 data Error
+  = NegativeSqrt Expr
+  | DivisionByZero Expr
 
 instance Show Error where
-  show = undefined
+  show (NegativeSqrt expr)   = "Error: Square root of a negative number: \"" ++ show expr ++ "\""
+  show (DivisionByZero expr) = "Error: Division by zero: \"" ++ show expr ++ "\""
 
 instance Eq Error where
-  (==) = undefined
+  (NegativeSqrt _)   == (NegativeSqrt _) = True
+  (DivisionByZero _) == (DivisionByZero _) = True
+  _ == _ = False
```

Comment:
I would argue that `NegativeSqrt (Num (-1))` and `NegativeSqrt (Add (Num 2) (Num 3))` are different errors

Reply:
[Fixed by using derivation](https://github.com/kajigor/fp-2024-cub-fall/pull/52#discussion_r1784555719)